File: t/VB/JPXDemo/UI/ViewerForm.vb
' *************************************************************
' Copyright (c) 1991-2019 LEAD Technologies, Inc.              
' All Rights Reserved.                                         
' *************************************************************
Imports Microsoft.VisualBasic
Imports System
Imports System.Drawing
Imports System.Collections
Imports System.ComponentModel
Imports System.Windows.Forms

Imports Leadtools
Imports Leadtools.Codecs
Imports Leadtools.Controls
Imports Leadtools.ImageProcessing
Imports System.Collections.Generic
Imports Leadtools.Jpeg2000
Imports Leadtools.ImageProcessing.Effects
Imports System.Drawing.Drawing2D
Imports Leadtools.Drawing

Public Class ViewerForm
   Inherits System.Windows.Forms.Form

   ''' <summary>
   ''' Required designer variable.
   ''' </summary>
   Private components As System.ComponentModel.Container = Nothing

   Public Sub New(ByVal parent As Form, ByVal bIsComposite As Boolean)
      '
      ' Required for Windows Form Designer support
      '
      InitializeComponent()

      '
      ' TODO: Add any constructor code after InitializeComponent call
      '
      MdiParent = parent
      IsComposite = bIsComposite
      InitClass()
   End Sub

   ''' <summary>
   ''' Clean up any resources being used.
   ''' </summary>
   Protected Overloads Overrides Sub Dispose(ByVal disposing As Boolean)
      If disposing Then
         If Not components Is Nothing Then
            components.Dispose()
         End If

         If Not ImageListControl Is Nothing Then
            ImageListControl.Items.Clear()
         End If

         If Not ImagesList Is Nothing Then
            ImagesList.Clear()
         End If

         If Not ColorList Is Nothing Then
            ColorList.Clear()
         End If

         If Not OpacityList Is Nothing Then
            OpacityList.Clear()
         End If

         If Not PreOpacityList Is Nothing Then
            PreOpacityList.Clear()
         End If


      End If
      MyBase.Dispose(disposing)
   End Sub

#Region "Windows Form Designer generated code"
   ''' <summary>
   ''' Required method for Designer support - do not modify
   ''' the contents of Me method with the code editor.
   ''' </summary>
   Private Sub InitializeComponent()
      Dim resources As System.ComponentModel.ComponentResourceManager = New System.ComponentModel.ComponentResourceManager(GetType(ViewerForm))
      Me._pnlViewer = New System.Windows.Forms.Panel()
      Me._pnlImageList = New System.Windows.Forms.Panel()
      Me.SuspendLayout()
      '
      '_pnlViewer
      '
      Me._pnlViewer.Anchor = CType((((System.Windows.Forms.AnchorStyles.Top Or System.Windows.Forms.AnchorStyles.Bottom) _
            Or System.Windows.Forms.AnchorStyles.Left) _
            Or System.Windows.Forms.AnchorStyles.Right), System.Windows.Forms.AnchorStyles)
      Me._pnlViewer.Location = New System.Drawing.Point(133, -2)
      Me._pnlViewer.Name = "_pnlViewer"
      Me._pnlViewer.Size = New System.Drawing.Size(679, 420)
      Me._pnlViewer.TabIndex = 3
      '
      '_pnlImageList
      '
      Me._pnlImageList.Anchor = CType(((System.Windows.Forms.AnchorStyles.Top Or System.Windows.Forms.AnchorStyles.Bottom) _
            Or System.Windows.Forms.AnchorStyles.Left), System.Windows.Forms.AnchorStyles)
      Me._pnlImageList.Location = New System.Drawing.Point(-1, -2)
      Me._pnlImageList.Name = "_pnlImageList"
      Me._pnlImageList.Size = New System.Drawing.Size(132, 421)
      Me._pnlImageList.TabIndex = 2
      '
      'ViewerForm
      '
      Me.AutoScaleBaseSize = New System.Drawing.Size(5, 13)
      Me.ClientSize = New System.Drawing.Size(810, 418)
      Me.Controls.Add(Me._pnlViewer)
      Me.Controls.Add(Me._pnlImageList)
      Me.Icon = CType(resources.GetObject("$this.Icon"), System.Drawing.Icon)
      Me.Name = "ViewerForm"
      Me.ShowInTaskbar = False
      Me.Text = "ViewerForm"
      Me.ResumeLayout(False)

   End Sub
#End Region

   Private _viewer As ImageViewer
   Private _imageListControl As ImageViewer

   Private _imagesList As List(Of RasterImage)
   Private _colorList As List(Of RasterImage)
   Private _opacityList As List(Of RasterImage)
   Private _preOpacityList As List(Of RasterImage)
   Private _name As String
   Private _activeList As ActiveImageLists
   Private _isComposite As Boolean
   Private _delay As Integer
   Private _renderWidth As Integer
   Private _renderHeight As Integer
   Private _playingAnnimation As Boolean
   Private _stopAnimation As Boolean
   Private _loopAnimation As Boolean

   Private Const DefaultViewerWidth As Integer = 600
   Private Const DefaultViewerHeight As Integer = 600
   Private WithEvents _pnlViewer As System.Windows.Forms.Panel
   Private WithEvents _pnlImageList As System.Windows.Forms.Panel
   Private Const DefaultDelay As Integer = 100

   Public Enum ActiveImageLists
      ImagesList = 0
      ColorList = 1
      OpacityList = 2
      PreOpacityList = 3
   End Enum

   Public Property LoopAnimation() As Boolean
      Get
         Return _loopAnimation
      End Get

      Set(ByVal value As Boolean)
         _loopAnimation = value
      End Set
   End Property

   Public Property StopAnimation() As Boolean
      Get
         Return _stopAnimation
      End Get
      Set(ByVal value As Boolean)
         _stopAnimation = value
      End Set
   End Property

   Public Property RenderHeight() As Integer
      Get
         Return _renderHeight
      End Get
      Set(ByVal value As Integer)
         _renderHeight = value
      End Set
   End Property

   Public Property RenderWidth() As Integer
      Get
         Return _renderWidth
      End Get
      Set(ByVal value As Integer)
         _renderWidth = value
      End Set
   End Property

   Public Property PlayingAnnimation() As Boolean
      Get
         Return _playingAnnimation
      End Get
      Set(ByVal value As Boolean)
         _playingAnnimation = value
      End Set
   End Property

   Public Property AnimationDelay() As Integer
      Get
         Return _delay
      End Get
      Set(ByVal value As Integer)
         _delay = value
      End Set
   End Property

   Public ReadOnly Property Image() As RasterImage
      Get
         Return _viewer.Image
      End Get
   End Property

   Public Property IsComposite() As Boolean
      Get
         Return _isComposite
      End Get
      Set(ByVal value As Boolean)
         _isComposite = value
      End Set
   End Property

   Public Property ImageListControl() As ImageViewer
      Get
         Return _imageListControl
      End Get
      Set(ByVal value As ImageViewer)
         _imageListControl = value
      End Set
   End Property

   Public Property ImagesList() As List(Of RasterImage)
      Get
         Return _imagesList
      End Get
      Set(ByVal value As List(Of RasterImage))
         _imagesList = value
      End Set
   End Property

   Public Property ActiveList() As ActiveImageLists
      Get
         Return _activeList
      End Get
      Set(ByVal value As ActiveImageLists)
         _activeList = value
      End Set
   End Property

   Public Property ColorList() As List(Of RasterImage)
      Get
         Return _colorList
      End Get
      Set(ByVal value As List(Of RasterImage))
         _colorList = value
      End Set
   End Property

   Private Property OpacityList() As List(Of RasterImage)
      Get
         Return _opacityList
      End Get

      Set(ByVal value As List(Of RasterImage))
         _opacityList = value
      End Set
   End Property

   Private Property PreOpacityList() As List(Of RasterImage)
      Get
         Return _preOpacityList
      End Get
      Set(ByVal value As List(Of RasterImage))
         _preOpacityList = value
      End Set
   End Property

   Public ReadOnly Property Viewer() As ImageViewer
      Get
         Return _viewer
      End Get
   End Property

   Private Sub InitClass()
      _viewer = New ImageViewer()
      _viewer.Dock = DockStyle.Fill
      _viewer.BorderStyle = BorderStyle.None
      AddHandler _viewer.TransformChanged, AddressOf _viewer_SizeModeChanged
      AddHandler _viewer.PostRender, AddressOf _viewer_PostTransformPaint
      _pnlViewer.Controls.Add(_viewer)
      _viewer.BringToFront()
      _viewer.AllowDrop = True
      _viewer.AutoDisposeImages = False

      '  Create a new RasterImageList control. 
      ImageListControl = New ImageViewer(New ImageViewerVerticalViewLayout() With {.Columns = 1})
      ImageListControl.SelectedItemBorderColor = Color.Red
      ImageListControl.SelectedItemBackgroundColor = Color.LightBlue
      ImageListControl.BackColor = Color.LightGray
      ImageListControl.BorderStyle = BorderStyle.FixedSingle
      ImageListControl.ImageBorderThickness = 1
      ImageListControl.ImageHorizontalAlignment = ControlAlignment.Far
      ImageListControl.ImageVerticalAlignment = ControlAlignment.Center
      ImageListControl.ItemBackgroundColor = Color.LightGray
      ImageListControl.ItemBorderThickness = 0
      ImageListControl.ItemPadding = New Padding(20)
      ImageListControl.ItemTextColor = Color.Black
      ImageListControl.ItemSize = LeadSize.Create(60, 80)
      ImageListControl.ItemSizeMode = ControlSizeMode.Fit
      ImageListControl.Dock = DockStyle.Fill
      Dim paintProperties As RasterPaintProperties
      paintProperties = ImageListControl.PaintProperties
      paintProperties.PaintDisplayMode = RasterPaintDisplayModeFlags.Bicubic
      ImageListControl.PaintProperties = paintProperties
      ImageListControl.Height = _viewer.Height
      ImageListControl.TextHorizontalAlignment = ControlAlignment.Center
      ImageListControl.TextVerticalAlignment = ControlAlignment.Center
      ImageListControl.InteractiveModes.BeginUpdate()
      ImageListControl.InteractiveModes.Add((New ImageViewerSelectItemsInteractiveMode() With {.SelectionMode = ImageViewerSelectionMode.Single}))
      ImageListControl.InteractiveModes.EndUpdate()
      AddHandler ImageListControl.SelectedItemsChanged, AddressOf ImageListControl_SelectedItemsChanged
      ImageListControl.AutoDisposeImages = False
      '  Add the RasterImageList to the control collection. 
      _pnlImageList.Controls.Add(ImageListControl)

      PlayingAnnimation = False
      StopAnimation = False
      LoopAnimation = False

      ActiveList = ActiveImageLists.ColorList

      ImagesList = New List(Of RasterImage)
      ColorList = New List(Of RasterImage)
      OpacityList = New List(Of RasterImage)
      PreOpacityList = New List(Of RasterImage)

      CType(MdiParent, MainForm).ClearCheck()
      CType(MdiParent, MainForm).SetCheck(ActiveList)

      RenderHeight = DefaultViewerHeight
      RenderWidth = DefaultViewerWidth
      AnimationDelay = DefaultDelay

      '  Add a handler to the PaintBackground event 
      ' ImageList.PaintBackground += new PaintEventHandler(rasterImageList_PaintBackground)
   End Sub

   Private Sub ImageListControl_SelectedItemsChanged(sender As Object, e As EventArgs)
      Dim items() As ImageViewerItem = ImageListControl.Items.GetSelected()
      If (items.Length = 0) Then
         _viewer.Image = Nothing
      Else
         ' Get the only item in the list since it is single selection
         Dim item As ImageViewerItem = items(0)
         Dim itemIndex As Integer = ImageListControl.Items.IndexOf(item)

         _viewer.Image = ImageListControl.Items(itemIndex).Image
      End If
      Viewer.Invalidate()
   End Sub


   Private Function LoadSaveAnimationStruct() As Boolean
      Dim _compositionBox As New CompositionBox()

      Try
         _compositionBox = CType(CType(MdiParent, MainForm).Jpeg2000Eng.ReadBox(_name, Jpeg2000BoxType.CompositionBox, 0), CompositionBox)
      Catch
         Return False
      End Try

      ' There must be only one composition box
      RenderHeight = _compositionBox.CompositionOptions.Height
      If (RenderHeight = 0) Then
         RenderHeight = DefaultViewerHeight
      End If

      RenderWidth = _compositionBox.CompositionOptions.Width
      If (RenderWidth = 0) Then
         RenderWidth = DefaultViewerWidth
      End If

      If ((Convert.ToInt32(_compositionBox.CompositionOptions.Loop) = 255)) Then
         LoopAnimation = True
      Else
         LoopAnimation = False
      End If

      AnimationDelay = 0

      If (_compositionBox.Instructions(0).Type = 4) Then
         AnimationDelay = _compositionBox.Instructions(0).Instructions(0).Life * _compositionBox.Instructions(0).Tick
      End If

      If (AnimationDelay = 0) Then
         AnimationDelay = DefaultDelay
      End If


      Return True
   End Function


   Private Function GetActiveList() As List(Of RasterImage)
      Select Case (ActiveList)
         Case ActiveImageLists.ColorList
            Return ColorList

         Case ActiveImageLists.OpacityList
            Return OpacityList

         Case ActiveImageLists.PreOpacityList
            Return PreOpacityList

         Case Else
            Return ImagesList
      End Select
   End Function

   Private Sub ImageList_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs)
      Viewer.Image = ImageListControl.Items.GetSelected()(0).Image

      Viewer.Image.Page = ImageListControl.Items.GetSelected()(0).PageNumber

      Viewer.Invalidate()
   End Sub

   Private Sub FillImageList(ByVal image As RasterImage)
      Using Codecs As New RasterCodecs()

         ImageListControl.Items.Clear()

         For index As Integer = 1 To image.PageCount
            image.Page = index
            '  Create the item of the image list 
            Dim item As New ImageViewerItem()
            item.Size = LeadSize.Create(120, 120)
            item.Text = "Page: " + (index - 1).ToString()
            item.Image = image.Clone()
            item.Tag = index

            ' Add the item to the image list 
            ImageListControl.Items.Add(item)
         Next index
         ImageListControl.Items(0).IsSelected = True
         _viewer.Image = ImageListControl.Items(0).Image
      End Using
   End Sub

   Public Sub FillImageList()
      Using codecs As New RasterCodecs()

         Dim activeList As List(Of RasterImage) = GetActiveList()

         If (IsNothing(activeList)) Then
            Messager.ShowError(Me, "No List Avaliable")
            Return
         End If

         Dim tempImage As New RasterImage(activeList(0))

         For index As Integer = 1 To activeList.Count - 1
            tempImage.AddPage(activeList(index).CloneAll())
         Next index

         FillImageList(tempImage)
      End Using
   End Sub

   Public Sub FillImages(ByVal compositeImage As List(Of CompositeJpxImages))
      For index As Integer = 0 To compositeImage.Count - 1

         ColorList.Add(compositeImage(index).ColorImage)

         If Not (IsNothing(compositeImage(index).OpacityImage)) Then
            OpacityList.Add(compositeImage(index).OpacityImage)
         Else
            OpacityList.Add(CType(MdiParent, MainForm).NoOpacityBitmap)
         End If

         If Not (IsNothing(compositeImage(index).PreOpacityImage)) Then
            PreOpacityList.Add(compositeImage(index).PreOpacityImage)
         Else
            PreOpacityList.Add(CType(MdiParent, MainForm).NoPreOpacityBitmap)
         End If

         If Not (IsNothing(compositeImage(index).OpacityImage)) Then
            Dim combineImage As New RasterImage(compositeImage(index).ColorImage)

            Dim Command As New CombineCommand(compositeImage(index).OpacityImage, _
                                                        New LeadRect(0, 0, compositeImage(index).OpacityImage.Width, compositeImage(index).OpacityImage.Height), _
                                                        New LeadPoint(0, 0), _
                                                        CombineCommandFlags.OperationMultiply)
            Command.Run(combineImage)

            ImagesList.Add(combineImage)
         Else
            ImagesList.Add(compositeImage(index).ColorImage)
         End If
      Next index
   End Sub

   Public Sub Initialize(ByVal info As ImageInformation, ByVal paintProperties As RasterPaintProperties, ByVal bSnap As Boolean, ByVal isfile As Boolean)
      _viewer.BeginUpdate()
      UpdatePaintProperties(paintProperties)
      FillImageList(info.Image)

      If Not (IsNothing(_viewer.Image)) Then
         AddHandler _viewer.Image.Changed, AddressOf Image_Changed
      End If
      _name = info.Name

      If (bSnap) Then
         Snap()
      End If
      If (isfile) Then
         LoadSaveAnimationStruct()
      End If
      UpdateCaption()
      _viewer.EndUpdate()
   End Sub

   Public Sub Initialize(ByVal compositeImage As List(Of CompositeJpxImages), ByVal fileName As String, ByVal paintProperties As RasterPaintProperties, ByVal bSnap As Boolean)
      _viewer.BeginUpdate()
      UpdatePaintProperties(paintProperties)

      ' _viewer.Image = compositeImage[0].ColorImage
      FillImages(compositeImage)
      FillImageList()

      If Not (IsNothing(_viewer.Image)) Then
         AddHandler _viewer.Image.Changed, AddressOf Image_Changed
      End If

      _name = fileName
      If (bSnap) Then
         Snap()
      End If

      LoadSaveAnimationStruct()
      UpdateCaption()
      _viewer.EndUpdate()
   End Sub

   Public Sub UpdatePaintProperties(ByVal paintProperties As RasterPaintProperties)
      _viewer.PaintProperties = paintProperties
   End Sub

   Private Sub UpdateCaption()
      Text = String.Format(_name)
   End Sub

   Private Sub Image_Changed(ByVal sender As Object, ByVal e As RasterImageChangedEventArgs)
      UpdateCaption()
      If (Not IsNothing(MdiParent) AndAlso Not PlayingAnnimation) Then
         CType(MdiParent, MainForm).UpdateControls()
      End If
   End Sub

   Private Sub _viewer_SizeModeChanged(ByVal sender As Object, ByVal e As EventArgs)
      CType(MdiParent, MainForm).UpdateControls()
   End Sub

   Public Sub Snap()
      _viewer.ClientSize = New Size(_pnlImageList.Width + DefaultViewerWidth, DefaultViewerHeight)
      ClientSize = _viewer.ClientSize
   End Sub

   Private Sub _viewer_PostTransformPaint(ByVal sender As Object, e As ImageViewerRenderEventArgs)
      Dim drawFont As Font = New Font("Arial", 9, FontStyle.Bold)
      Try
         Dim g As Graphics = e.PaintEventArgs.Graphics
         Dim mm As LeadMatrix = _viewer.ViewTransform
         Dim m As Matrix = New Matrix(CSng(mm.M11), CSng(mm.M12), CSng(mm.M21), CSng(mm.M22), CSng(mm.OffsetX), CSng(mm.OffsetY))
         Dim t As Transformer = New Transformer(m)
      Finally
         drawFont.Dispose()
      End Try
   End Sub
End Class

File: t/CS/JpxDemo/UI/ViewerForm.cs
// *************************************************************
// Copyright (c) 1991-2019 LEAD Technologies, Inc.              
// All Rights Reserved.                                         
// *************************************************************
using System;
using System.Collections;
using System.Collections.Generic;
using System.ComponentModel;
using System.Drawing;
using System.Reflection;
using System.Windows.Forms;

using Leadtools;
using Leadtools.Codecs;
using Leadtools.Controls;
using Leadtools.Demos;
using Leadtools.Drawing;
using Leadtools.ImageProcessing;
using Leadtools.ImageProcessing.Effects;
using Leadtools.Jpeg2000;

namespace JPXDemo
{
   /// <summary>
   /// Summary description for ViewerForm.
   /// </summary>
   public class ViewerForm : System.Windows.Forms.Form
   {
      /// <summary>
      /// Required designer variable.
      /// </summary>
      private System.ComponentModel.Container components = null;

      public ViewerForm(Form parent, bool isComposite)
      {
         //
         // Required for Windows Form Designer support
         //
         InitializeComponent();

         //
         // TODO: Add any constructor code after InitializeComponent call
         //
         MdiParent = parent;
         IsComposite = isComposite;
         InitClass();
      }

      /// <summary>
      /// Clean up any resources being used.
      /// </summary>
      protected override void Dispose(bool disposing)
      {
         if(disposing)
         {
            if(components != null)
            {
               components.Dispose();
            }

            if(ImageListControl != null)
               ImageListControl.Items.Clear();

            if (ImagesList != null)
               ImagesList.Clear();

            if (ColorList != null)
               ColorList.Clear();

            if(OpacityList != null)
               OpacityList.Clear();

            if(PreOpacityList != null)
               PreOpacityList.Clear();
         }
         base.Dispose(disposing);
      }

      #region Windows Form Designer generated code
      /// <summary>
      /// Required method for Designer support - do not modify
      /// the contents of this method with the code editor.
      /// </summary>
      private void InitializeComponent( )
      {
         System.ComponentModel.ComponentResourceManager resources = new System.ComponentModel.ComponentResourceManager(typeof(ViewerForm));
         this._pnlImageList = new System.Windows.Forms.Panel();
         this._pnlViewer = new System.Windows.Forms.Panel();
         this.SuspendLayout();
         // 
         // _pnlImageList
         // 
         this._pnlImageList.Anchor = ((System.Windows.Forms.AnchorStyles)(((System.Windows.Forms.AnchorStyles.Top | System.Windows.Forms.AnchorStyles.Bottom) 
            | System.Windows.Forms.AnchorStyles.Left)));
         this._pnlImageList.Location = new System.Drawing.Point(0, 1);
         this._pnlImageList.Name = "_pnlImageList";
         this._pnlImageList.Size = new System.Drawing.Size(128, 376);
         this._pnlImageList.TabIndex = 0;
         // 
         // _pnlViewer
         // 
         this._pnlViewer.Anchor = ((System.Windows.Forms.AnchorStyles)((((System.Windows.Forms.AnchorStyles.Top | System.Windows.Forms.AnchorStyles.Bottom) 
            | System.Windows.Forms.AnchorStyles.Left) 
            | System.Windows.Forms.AnchorStyles.Right)));
         this._pnlViewer.Location = new System.Drawing.Point(134, 1);
         this._pnlViewer.Name = "_pnlViewer";
         this._pnlViewer.Size = new System.Drawing.Size(678, 375);
         this._pnlViewer.TabIndex = 1;
         // 
         // ViewerForm
         // 
         this.AutoScaleBaseSize = new System.Drawing.Size(5, 13);
         this.ClientSize = new System.Drawing.Size(809, 373);
         this.Controls.Add(this._pnlViewer);
         this.Controls.Add(this._pnlImageList);
         this.Icon = ((System.Drawing.Icon)(resources.GetObject("$this.Icon")));
         this.Name = "ViewerForm";
         this.ShowInTaskbar = false;
         this.Text = "ViewerForm";
         this.Closed += new System.EventHandler(this.ViewerForm_Closed);
         this.Deactivate += new System.EventHandler(this.ViewerForm_Deactivate);
         this.ResumeLayout(false);

      }
      #endregion

      private ImageViewer _viewer;
      private ImageViewer _imageListControl;

      private List<RasterImage> _imagesList;
      private List<RasterImage> _colorList;
      private List<RasterImage> _opacityList;
      private List<RasterImage> _preOpacityList;
      private Panel _pnlImageList;
      private Panel _pnlViewer;
      private string _name;
      private ActiveImageLists _activeList;
      private bool _isComposite;
      private int _delay;
      private int _renderWidth;
      private int _renderHeight;
      private bool _playingAnnimation;
      private bool _stopAnimation;
      private bool _loopAnimation;

      private const int DefaultViewerWidth = 600;
      private const int DefaultViewerHeight = 600;
      private const int DefaultDelay = 100;

      public enum ActiveImageLists
      {
         ImagesList = 0,
         ColorList = 1,
         OpacityList = 2,
         PreOpacityList = 3,
      }

      public bool LoopAnimation
      {
         get
         {
            return _loopAnimation;
         }
         set
         {
            _loopAnimation = value;
         }
      }

      public bool StopAnimation
      {
         get
         {
            return _stopAnimation;
         }
         set
         {
            _stopAnimation = value;
         }
      }

      public int RenderHeight
      {
         get
         {
            return _renderHeight;
         }
         set
         {
            _renderHeight = value;
         }
      }

      public int RenderWidth
      {
         get
         {
            return _renderWidth;
         }
         set
         {
            _renderWidth = value;
         }
      }
      public bool PlayingAnnimation
      {
         get
         {
            return _playingAnnimation;
         }
         set
         {
            _playingAnnimation = value;
         }
      }

      public int AnimationDelay
      {
         get
         {
            return _delay;
         }
         set
         {
            _delay = value;
         }
      }

      public RasterImage Image
      {
         get
         {
            return _viewer.Image;
         }
      }

      public bool IsComposite
      {
         get
         {
            return _isComposite;
         }
         set
         {
            _isComposite = value;
         }
      }

      public ImageViewer Viewer
      {
         get
         {
            return _viewer;
         }
      }

      public ImageViewer ImageListControl
      {
         get
         {
            return _imageListControl;
         }
         set
         {
            _imageListControl = value;
         }
      }

      public List<RasterImage> ImagesList
      {
         get
         {
            return _imagesList;
         }
         set
         {
            _imagesList = value;
         }
      }

      public ActiveImageLists ActiveList
      {
         get
         {
            return _activeList;
         }
         set
         {
            _activeList = value;
         }
      }

      public List<RasterImage> ColorList
      {
         get
         {
            return _colorList;
         }
         set
         {
            _colorList = value;
         }
      }

      private List<RasterImage> OpacityList
      {
         get
         {
            return _opacityList;
         }
         set
         {
            _opacityList = value;
         }
      }

      private List<RasterImage> PreOpacityList
      {
         get
         {
            return _preOpacityList;
         }
         set
         {
            _preOpacityList = value;
         }
      }

      private void InitClass( )
      {
         _viewer = new ImageViewer();
         _viewer.Dock = DockStyle.Fill;
         _viewer.BorderStyle = BorderStyle.None;
         _pnlViewer.Controls.Add(_viewer);
         _viewer.BringToFront();
         _viewer.AllowDrop = true;
         _viewer.AutoDisposeImages = false;

         // Create a new RasterImageList control. 
         ImageViewerVerticalViewLayout viewLayout = new Leadtools.Controls.ImageViewerVerticalViewLayout() { Columns = 1 };
         ImageListControl = new ImageViewer(viewLayout);
         Size = new Size(200, 200);
         ImageListControl.Bounds = new Rectangle(new Point(0, 0), Size);

         ImageListControl.SelectedItemBorderColor = Color.Red;
         ImageListControl.SelectedItemBackgroundColor = Color.LightBlue;
         ImageListControl.BackColor = Color.LightGray;
         ImageListControl.BorderStyle = BorderStyle.FixedSingle;
         ImageListControl.ImageBorderThickness = 1;
         ImageListControl.ImageHorizontalAlignment = ControlAlignment.Far;
         ImageListControl.ImageVerticalAlignment = ControlAlignment.Center;
         ImageListControl.ItemBackgroundColor = Color.LightGray;
         ImageListControl.ItemBorderThickness = 0;
         ImageListControl.ItemPadding = new Padding(20);
         ImageListControl.ItemTextColor = Color.Black;
         ImageListControl.ItemSize = LeadSize.Create(60, 80);
         ImageListControl.ItemSizeMode = ControlSizeMode.Fit;
         ImageListControl.Dock = DockStyle.Fill;
         RasterPaintProperties paintProperties = ImageListControl.PaintProperties;
         paintProperties.PaintDisplayMode = RasterPaintDisplayModeFlags.Bicubic;
         ImageListControl.PaintProperties = paintProperties;
         ImageListControl.Height = _viewer.Height;
         ImageListControl.TextHorizontalAlignment = ControlAlignment.Center;
         ImageListControl.TextVerticalAlignment = ControlAlignment.Center;
         ImageListControl.InteractiveModes.BeginUpdate();
         ImageListControl.InteractiveModes.Add(new ImageViewerSelectItemsInteractiveMode() { SelectionMode = ImageViewerSelectionMode.Single });
         ImageListControl.InteractiveModes.EndUpdate();
         ImageListControl.SelectedItemsChanged += new EventHandler(ImageListControl_SelectedItemsChanged);
         ImageListControl.AutoDisposeImages = false;

         // Add the RasterImageList to the control collection. 
         _pnlImageList.Controls.Add(ImageListControl);

         PlayingAnnimation = false;
         StopAnimation = false;
         LoopAnimation = false;
         
         ActiveList = ActiveImageLists.ColorList;

         ImagesList = new List<RasterImage>();
         ColorList = new List<RasterImage>();
         OpacityList = new List<RasterImage>();
         PreOpacityList = new List<RasterImage>();

         ((MainForm)MdiParent).ClearCheck();
         ((MainForm)MdiParent).SetCheck(ActiveList);

         RenderHeight = DefaultViewerHeight;
         RenderWidth = DefaultViewerWidth;
         AnimationDelay = DefaultDelay;

      }      

      private bool LoadSaveAnimationStruct()
      {
         CompositionBox _compositionBox = new CompositionBox();

         try
         {
            _compositionBox = (CompositionBox) ((MainForm)MdiParent).Jpeg2000Eng.ReadBox(_name, Jpeg2000BoxType.CompositionBox, 0);
         }
         catch
         {
            return false;
         }

         // There must be only one composition box
         RenderHeight = _compositionBox.CompositionOptions.Height;
         if(RenderHeight == 0)
            RenderHeight = DefaultViewerHeight;

         RenderWidth = _compositionBox.CompositionOptions.Width;
         if (RenderWidth == 0)
            RenderWidth = DefaultViewerWidth;

         LoopAnimation = (Convert.ToInt32(_compositionBox.CompositionOptions.Loop) == 255) ? true : false;
         AnimationDelay = 0;

         if (_compositionBox.Instructions[0].Type == 4)
            AnimationDelay = _compositionBox.Instructions[0].Instructions[0].Life * _compositionBox.Instructions[0].Tick;

         if (AnimationDelay == 0)
            AnimationDelay = DefaultDelay;

         return true;
      }

      private List<RasterImage> GetActiveList()
      {
         switch (ActiveList)
         {
            case ActiveImageLists.ColorList:
               return ColorList;

            case ActiveImageLists.OpacityList:
               return OpacityList;

            case ActiveImageLists.PreOpacityList:
               return PreOpacityList;

            default:
               return ImagesList;
         }
      }

      void ImageListControl_SelectedItemsChanged(object sender, EventArgs e)
      {
         var items = ImageListControl.Items.GetSelected();
         if (items.Length == 0)
         {
            _viewer.Image = null;
         }
         else
         {
            //Get the only item in the list since it is single selection
            var item = items[0];
            var itemIndex = ImageListControl.Items.IndexOf(item);
            
            _viewer.Image = ImageListControl.Items[itemIndex].Image;

         }

         Viewer.Invalidate();
      }

      private void FillImageList(RasterImage image)
      {
         using(RasterCodecs codecs = new RasterCodecs())
         {
            ImageListControl.Items.Clear();

            for(int index = 1; index <= image.PageCount; index++)
            {
               image.Page = index;
               // Create the item of the image list 
               ImageViewerItem item = new ImageViewerItem();
               item.Size = LeadSize.Create(120, 120);
               item.Text = "Page: " + (index - 1).ToString();
               item.Image = image.Clone();
               item.Tag = index;

               // Add the item to the image list 
               ImageListControl.Items.Add(item);
            }

            ImageListControl.Items[0].IsSelected = true;
            _viewer.Image = ImageListControl.Items[0].Image;
         }
      }

      public void FillImageList()
      {
         using(RasterCodecs codecs = new RasterCodecs())
         {
            List<RasterImage> activeList = GetActiveList();

            if(activeList == null)
            {
               Messager.ShowError(this, "No List Avaliable");
               return;
            }

            RasterImage tempImage = new RasterImage(activeList[0]);
            for(int index = 1; index < activeList.Count; index++)
            {
               tempImage.AddPage(activeList[index].CloneAll());
            }
            FillImageList(tempImage);
         }
      }

      public void FillImages(List<CompositeJpxImages> compositeImage)
      {
         for (int index = 0; index < compositeImage.Count; index++)
         {
            ColorList.Add(compositeImage[index].ColorImage);

            if (compositeImage[index].OpacityImage != null)
               OpacityList.Add(compositeImage[index].OpacityImage);
            else
               OpacityList.Add(((MainForm)MdiParent).NoOpacityBitmap);

            if (compositeImage[index].PreOpacityImage != null)
               PreOpacityList.Add(compositeImage[index].PreOpacityImage);
            else
               PreOpacityList.Add(((MainForm)MdiParent).NoPreOpacityBitmap);

            if (compositeImage[index].OpacityImage != null)
            {
               RasterImage combineImage = new RasterImage(compositeImage[index].ColorImage);

               CombineCommand command = new CombineCommand(compositeImage[index].OpacityImage,
                                                           new LeadRect(0, 0, compositeImage[index].OpacityImage.Width, compositeImage[index].OpacityImage.Height),
                                                           new LeadPoint(0, 0),
                                                           CombineCommandFlags.OperationMultiply);
               command.Run(combineImage);

               ImagesList.Add(combineImage);
            }
            else
            {
               ImagesList.Add(compositeImage[index].ColorImage);
            }
         }
      }

      public void Initialize(ImageInformation info, RasterPaintProperties paintProperties, bool snap,bool isFile)
      {
         _viewer.BeginUpdate();
         UpdatePaintProperties(paintProperties);
         FillImageList(info.Image);
         if (_viewer.Image != null)
            _viewer.Image.Changed += new EventHandler<RasterImageChangedEventArgs>(Image_Changed);
         _name = info.Name;
         if(snap)
            Snap();
         if(isFile)
            LoadSaveAnimationStruct();
         UpdateCaption();
         _viewer.EndUpdate();
      }

      public void Initialize(List<CompositeJpxImages> compositeImage, string fileName, RasterPaintProperties paintProperties, bool snap)
      {
         _viewer.BeginUpdate();
         UpdatePaintProperties(paintProperties);
         FillImages(compositeImage);
         FillImageList();
         if (_viewer.Image != null)
            _viewer.Image.Changed += new EventHandler<RasterImageChangedEventArgs>(Image_Changed);
         _name = fileName;
         if (snap)
            Snap();
         LoadSaveAnimationStruct();
         UpdateCaption();
         _viewer.EndUpdate();
      }

      public void UpdatePaintProperties(RasterPaintProperties paintProperties)
      {
         _viewer.PaintProperties = paintProperties;
      }

      private void UpdateCaption( )
      {
         Text = string.Format(_name);
      }

      private void Image_Changed(object sender, RasterImageChangedEventArgs e)
      {
         UpdateCaption();
         if(MdiParent != null && !PlayingAnnimation)
            ((MainForm)MdiParent).UpdateControls();
      }

      private void _viewer_SizeModeChanged(object sender, EventArgs e)
      {
         ((MainForm)MdiParent).UpdateControls();
      }

      public void Snap( )
      {
         _viewer.ClientSize = new Size(_pnlImageList.Width + DefaultViewerWidth, DefaultViewerHeight);
         ClientSize = _viewer.ClientSize;
      }

      private void ViewerForm_Closed(object sender, System.EventArgs e)
      {
      }

      private void ViewerForm_Deactivate(object sender, System.EventArgs e)
      {
      }
   }
}
